arangod/Metrics/ClusterMetricsFeature.cpp

```diff
@@ -0,0 +1,192 @@
+#include "Statistics/StatisticsFeature.h"
+#include "StorageEngine/EngineSelectorFeature.h"
+
+namespace arangodb::metrics {
+using namespace std::chrono_literals;
+
+ClusterMetricsFeature::ClusterMetricsFeature(Server& server)
+    : ArangodFeature{server, *this} {
+  setOptional();
+  startsAfter<ClusterFeature>();
+  startsAfter<NetworkFeature>();
+  startsAfter<SchedulerFeature>();
+}
+
+void ClusterMetricsFeature::collectOptions(
+    std::shared_ptr<options::ProgramOptions> options) {
+  options
+      ->addOption(
+          "--server.cluster-metrics-timeout",
+          "Seconds that will be between cross cluster calls that collect "
+          "metrics",
+          new options::UInt32Parameter(&_timeout),
+          arangodb::options::makeDefaultFlags(arangodb::options::Flags::Hidden))
+      .setIntroducedIn(310000);
+}
+
+void ClusterMetricsFeature::validateOptions(
+    std::shared_ptr<options::ProgramOptions> /*options*/) {
+  if (!ServerState::instance()->isCoordinator()) {
+    disable();
+  }
+}
+void ClusterMetricsFeature::beginShutdown() {
+  _count.store(std::numeric_limits<int32_t>::min() + 1,
+               std::memory_order_relaxed);
+}
+
+void ClusterMetricsFeature::asyncUpdate() {
+  if (isEnabled() && _count.fetch_add(1, std::memory_order_acq_rel) == 0) {
+    scheduleUpdate();
+  }
+}
+
+void ClusterMetricsFeature::scheduleUpdate() noexcept {
+  auto now = std::chrono::steady_clock::now();
+  auto next = _lastUpdate + std::chrono::seconds{_timeout};
+  _handle = SchedulerFeature::SCHEDULER->queueDelayed(
+      RequestLane::CLUSTER_INTERNAL, (now < next ? next - now : 0ns),
+      [this](bool canceled) {
+        if (canceled) {
+          return;
+        }
+        _count.store(1, std::memory_order_relaxed);
+        try {
+          update();
+        } catch (...) {
+          repeatUpdate();
+        }
+      });
+}
+
+void ClusterMetricsFeature::update() {
+  auto& nf = server().getFeature<NetworkFeature>();
+  auto& cf = server().getFeature<ClusterFeature>();
+  // TODO try get from agency
+  metricsOnCoordinator(nf, cf).thenFinal(
+      [this](futures::Try<RawDBServers>&& metrics) mutable {
+        // We want more time than kTimeout should expire
+        // after last try cross cluster communication
+        _lastUpdate = std::chrono::steady_clock::now();
+        if (!metrics.hasValue()) {
+          scheduleUpdate();
+          return;
+        }
+        try {
+          set(std::move(*metrics));
+        } catch (...) {
+        }
+        repeatUpdate();
+      });
+}
+
+void ClusterMetricsFeature::repeatUpdate() noexcept {
+  if (_count.fetch_sub(1, std::memory_order_acq_rel) > 1) {
+    scheduleUpdate();
+  }
+}
+
+void ClusterMetricsFeature::set(RawDBServers&& raw) const {
+  Metrics metrics;
+  std::shared_lock lock{_m};
+  for (auto const& payload : raw) {
+    TRI_ASSERT(payload);
+    velocypack::Slice slice{payload->data()};
+    auto const size = slice.length();
+    TRI_ASSERT(size % 3 == 0);
```

Comment:
* Check slice type
* Turn assertion into real check of array size

Reply:
> Turn assertion into real check of array size

Do we really want this? I don't like unreachable code